### ISSUE TITLE
DefaultRecoverable doesn't seem to deal with nodes going out-to-lunch and then returning

### ISSUE THREAD

**Issue by chetmurthy** [bft-smart/library#27]
I've written a simple test application with BFT-Smart, as a warmup for doing something real.  This application consists in:

(a) a simple client that sends "commands" consisting of some fixed amount of random bytes (1kbytes in this case).  Perhaps one command, perhaps 100k commands serially, depending on the test

(b) The server-side application starts off with a SHA1 hash that's all zeroes, and for each command, appends the hash and the command, takes the hash of it, and sets that to be the new hash.

  --> So: a primitive sort of "state" that can be affected by a large "command".

=============

I'm running with 4 replicas, and the server extends DefaultRecoverable.

If I start 4 replicas, fire up a client doing 100k writes, and then after some thousands of writes, "control-Z" (stop, not kill) one replica, the other replicas keep going.  MUCH, much later, if I "fg" (resume) the stopped replica, it attempts to catch up, and many, many bad things start happening:

(i) initially, {in,out}QueueSize was 500k msgs, and I got OutOfMemory errors.  I reduced those to 5000 messages.

(ii) Then the stopped/resumed replica gets errors b/c it's run out of buffers, so it discards messages, and then asks for state-transfer.  This state-transfer never successfully completes.

(iii) Eventually, even in this is much-more-constrained scenario, two replicas end up suffering OOM errors (NOT the stopped/started replica) and I killed the test off.

I'm not sure how to go about debugging this, and/or how to go about providing enough detail that you can reproduce it.  I'm happy to share code and steps to repro, if you'd like.

Also, is the sort of faults I induced supposed to be tolerated by 4-replica DefaultRecoverable ?

Thanks,
--chet--

**Comment by chetmurthy** [comment]
I should have added a little more:

(iv) in at least a few instances, the stopped/resumed replica seems to fire up a number of threads that go into some sort of lock-spin, taking enormous CPU.  I didn't take a javacore (again: wanted to first get advice on whether/how to go about debugging)

(v) I further reduced the QueueSize to 1k msgs.  At this point, I don't get OOMs, but the stopped/started replica never gets reconnected/up-to-date.

Is there any way to get a status-dump from replicas, that one could compare?  For intsance, right now, I'm sending no traffic to the cluster, so I could take a status-dump (howsoever exhaustive) and (if I knew what I was doing) figure out why the stopped/resumed replica seems to have given up trying to get back in sync.

**Comment by snakejerusalem** [comment]
Greetings,

Are you using the version available for download (v1.1-beta, released in last December) or have you fetched the code available in the master branch? If you are using v1.1-beta, try fetching the code available at the master branch of the repository. We have recently patched the code to fix a bug that is pretty similar to what you have described here: if the client issued large requests, a replica would be unable to properly de-serialize the logs sent from the other ones.

**Comment by snakejerusalem** [comment]
Also, does this problem still arises if you ctrl + c the replica and re-start it?

**Comment by chetmurthy** [comment]
Hi, thank you for such a quick reply!

(1) I hadn't tried "master", but am doing so right now.  I did 1 read, then 100k writes, and then left things sit for a little while; when I came back to look, the system seemed to be stuck, and I could not do a single write or read.

(1') I ran the same test again, this time, stopping the writes at around 150k (killed client).  one of the replicas seems permanently stuck behind, giving wrong answers to reads.  Every replica is printing log messages of the form:

(SendStopTask.run) Re-transmitting STOP message to install regency 1

and the replica that is stuck is (indeed) replica 1

(2) I did control-C on the lagging replica and restarted it a few times with 1.1-beta.  State-transfer blew up badly, though I don't remember the details.

The load scenario I'm running is pretty "friendly".  A single client, submitting 1kbyte requests (and waiting for a reply to each) as fast as it can.  Can you suggest a scenario that should work, from which I could start?

**Comment by chetmurthy** [comment]
I re-ran "4 replicas, client doing 1m writes", then control-C replica 3.  eventually, the other replicas get reformed and start moving (the timeout seems a bit long).  I then restarted replica 3.  After a lot of noisy logging, I can see that replica 3 does not get back up-to-date, and seems permanently stuck.

A read confirms that replica 3 is far, far behind.

So control-C/restart doesn't seem to work either.

**Comment by snakejerusalem** [comment]
Some of the behaviour you describe seem to be consistent with a huge consumption of memory (which is prone to happen in DefaultRecoverable, since it never stores logs or state to disk).

You said that each replica state is generated by 1) starting with an hash set to all-zeros 2) append to each command the hash and itself, generate the hash of it all, then set the result to be the new hash. I reckon this is made to keep each stored command at a fixed length, but do you also keep a fixed amount of commands? If the number of commands stored as the application state is growing, that might be the cause for the system to become so unstable.

Also, did you modified the "system.totalordermulticast.checkpoint_period" to a larger value? If so, that might also be consuming memory. Have you also tried to increased the quantity of memory allocated to the JVM?

**Comment by chetmurthy** [comment]
(1) your description of this app is clear.  It's pretty trivial.  I don't save commands -- just the consensus-id and the SHA256.

(2) Here are the changes I made to system.config.  I'm down to 1k Queue-size.  Maybe I should be -decreasing- the checkpoint period?

--- a/bft-smart-apps/config/system.config
+++ b/bft-smart-apps/config/system.config
@@ -49,10 +49,12 @@ system.totalordermulticast.nonces = 10
 system.totalordermulticast.verifyTimestamps = false
 
 #Quantity of messages that can be stored in the receive queue of the communication system
-system.communication.inQueueSize = 500000
+#system.communication.inQueueSize = 500000
+system.communication.inQueueSize = 1000
 
 # Quantity of messages that can be stored in the send queue of each replica
-system.communication.outQueueSize = 500000
+#system.communication.outQueueSize = 500000
+system.communication.outQueueSize = 1000
 
 #Set to 1 if SMaRt should use signatures, set to 0 if otherwise
 system.communication.useSignatures = 0

**Comment by chetmurthy** [comment]
Oops.  Also, since I'm no longer suffering OOMs (with 1k queue-size) I'm running with default JVM heap.  Previously, I had set the heapsize as large as 1Gig and still seen errors.  But of course, with 500k msgs queue-size and 1kbyte message, that's completely reasonable *grin*

**Comment by snakejerusalem** [comment]
(1) Yes, but do you save a single consensus-id/SHA pair, or do you save an unbounded list of consensus-id/SHA pairs?

(2) If you are issuing commands with 1 kbyte, they will be stored in a log that will periodically get flushed (and then a snapshot of the state is taken). A period of 1000 is ought to be enough even for 1kb commands, but if the state itself grows akin what I asked in (1), you will run into memory problems.

**Comment by chetmurthy** [comment]
Re: (1): here's my "State" class (and the DefaultRecoverable class has a single instance-var of this class):


    class State {
	public int commit_id ;
	public byte[] chained_hash;
	State() {
	    commit_id = 0 ;
	    chained_hash = new byte[32];
	}

	public State copy() {
	    State st2 = new State();
	    st2.commit_id = commit_id ;
	    st2.chained_hash = chained_hash.clone();
	    return st2;
	}

	public byte[] serialize() {
	    ByteArrayOutputStream out = new ByteArrayOutputStream();
	    DataOutputStream dos = new DataOutputStream(out);
	    try {
		dos.writeInt(commit_id) ;
		dos.write(chained_hash, 0, chained_hash.length) ;
	    }
	    catch (IOException e) {
		System.out.println("Exception writing data in the replica: " + e.getMessage());
		e.printStackTrace();
		System.exit(-1);
	    }
	    byte[] b = out.toByteArray() ;
	    assert (b.length == 44) ;
	    return b;
	}

	public void deserialize(byte[] state) {
	    assert (state.length == 36) ;
	    assert (chained_hash.length == 32) ;
	    ByteArrayInputStream in = new ByteArrayInputStream(state);
	    DataInputStream dis = new DataInputStream(in);
	    try {
		commit_id = dis.readInt();
		dis.read(chained_hash, 0, chained_hash.length);
	    }
	    catch (IOException e) {
		System.out.println("Exception writing data in the replica: " + e.getMessage());
		e.printStackTrace();
		System.exit(-1);
	    }
	}	

	public byte[] executeSingle(byte[] command, MessageContext msgCtx) {
	    if (msgCtx.getConsensusId() != -1) {
		commit_id = msgCtx.getConsensusId() ;
		try {
		    chained_hash = AeSimpleSHA256.Hash(new byte[][]{ chained_hash, command });
		} catch (NoSuchAlgorithmException e) { 
		    // TODO Auto-generated catch block 
		    e.printStackTrace();
		    System.exit(-1);
		} catch (UnsupportedEncodingException e) { 
		    // TODO Auto-generated catch block 
		    e.printStackTrace();
		    System.exit(-1);
		}
	    }
	    return serialize();
	}
    }

**Comment by snakejerusalem** [comment]
Ok, the state size is clearly constrained.

I think I have an idea about why the delayed replica can't catch up to the others. You said you are running your app in a cluster, so the replicas are ought to be able to process requests quite efficiently, and when the delayed/crashed replica asks for the state with a log up to consensus X, the other replicas already already performed a checkpoint and discarded the log that the other replica was asking for (which is not a uncommon phenomenom). Try to decrease the rate at which the clients issue requests and/or increase the "system.totalordermulticast.checkpoint_period" parameter in config/system.config to a larger value. 

Also, keep using the code available in the master repository, because it already has a couple of patches that render it more stable than the latest release.

**Comment by chetmurthy** [comment]
Hi, can you tell me what the checkpoint_period is measured in?  milliseconds?  Seconds?

Also, from your suggestion, I conclude that DefaultRecoverable implements no state-transfer?  I would expect that a replica that had fallen too far behind, would be brought up-to-date by state-transfer, so the fact that other replicas had discarded logs would not be a problem?

I'm currently on the tip of master, and I'll stay there *grin*.

**Comment by snakejerusalem** [comment]
The checkpoint_period is not time units; the value represents a number of client commands. If the parameter is set to 1000, then a checkpoint will occur every 1000 commands.

DefaultRecoverable is an abstract class; you have to implement the getSnapshot() and setSnapshot(byte[] param) methods. The idea is to use getSnapshot() to pass to the library the current state serialized into a byte array (which is periodically invoked by all replicas, depending on the value described above). The log is discarted at this point, since the replica obtains a snapshot of the state. On the other and, setSnapshot(...) is invoked by the replica that is trying to catch up, to be able to install the state sent by the other replicas.

Also, maybe this example might help you out to understand how to implement DefaultRecoverable's abstract methods: https://github.com/bft-smart/library/wiki/Getting-Started-with-BFT-SMaRt

**Comment by chetmurthy** [comment]
Indeed, I've implemented the get/install snapshot operations (code included below -- wish I could attach a Java file, feh).  So I still don't understand why state-xfer isn't happening.  I actually started with the "foo" example at that link, and converted it into the test application I'm working with.  In the code below, there's an instance-variable "ob", which is (for the tests I'm running now) null.  Eventually, it'll point at an outboard server to which operations will be forwarded, but right now, I'm focused on getting "inboard" operation working.


package ob.server;

// These are the classes which receive requests from clients
import bftsmart.reconfiguration.util.HostsConfig;
import bftsmart.tom.MessageContext;
import bftsmart.tom.ServiceReplica;
import bftsmart.tom.server.defaultservices.DefaultRecoverable;
import bftsmart.tom.server.defaultservices.durability.DurabilityCoordinator;


// Classes that need to be declared to implement this
// replicated Map
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.UnsupportedEncodingException; 
import java.security.NoSuchAlgorithmException; 

import java.nio.ByteBuffer;
import java.util.List;
import java.util.ArrayList;

import hash.AeSimpleSHA256;

import org.apache.thrift.TException;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;

import ob.thrift.*;

public class OBProxy extends /*DurabilityCoordinator*/ DefaultRecoverable {

    class State {
	public int commit_id ;
	public byte[] chained_hash;
	State() {
	    commit_id = 0 ;
	    chained_hash = new byte[32];
	}

	public State copy() {
	    State st2 = new State();
	    st2.commit_id = commit_id ;
	    st2.chained_hash = chained_hash.clone();
	    return st2;
	}

	public byte[] serialize() {
	    ByteArrayOutputStream out = new ByteArrayOutputStream();
	    DataOutputStream dos = new DataOutputStream(out);
	    try {
		dos.writeInt(commit_id) ;
		dos.write(chained_hash, 0, chained_hash.length) ;
	    }
	    catch (IOException e) {
		System.out.println("Exception writing data in the replica: " + e.getMessage());
		e.printStackTrace();
		System.exit(-1);
	    }
	    byte[] b = out.toByteArray() ;
	    assert (b.length == 44) ;
	    return b;
	}

	public void deserialize(byte[] state) {
	    assert (state.length == 36) ;
	    assert (chained_hash.length == 32) ;
	    ByteArrayInputStream in = new ByteArrayInputStream(state);
	    DataInputStream dis = new DataInputStream(in);
	    try {
		commit_id = dis.readInt();
		dis.read(chained_hash, 0, chained_hash.length);
	    }
	    catch (IOException e) {
		System.out.println("Exception writing data in the replica: " + e.getMessage());
		e.printStackTrace();
		System.exit(-1);
	    }
	}	

	public byte[] executeSingle(byte[] command, MessageContext msgCtx) {
	    if (msgCtx.getConsensusId() != -1) {
		commit_id = msgCtx.getConsensusId() ;
		try {
		    chained_hash = AeSimpleSHA256.Hash(new byte[][]{ chained_hash, command });
		} catch (NoSuchAlgorithmException e) { 
		    // TODO Auto-generated catch block 
		    e.printStackTrace();
		    System.exit(-1);
		} catch (UnsupportedEncodingException e) { 
		    // TODO Auto-generated catch block 
		    e.printStackTrace();
		    System.exit(-1);
		}
	    }
	    return serialize();
	}
    }

    State st ;
    Outboard.Client ob;

    private Outboard.Client connectOutboard(String host, int port) {
	try {
	    TTransport transport = new TSocket(host, port);
	    transport.open();

	    TProtocol protocol = new  TBinaryProtocol(transport);
	    Outboard.Client client = new Outboard.Client(protocol);
	    return client;
	} catch (TException x) {
	    x.printStackTrace();
	    System.exit(-1);
	    return null;
	}
    }
    public OBProxy(int id, String outboard_host, int outboard_port) {
	st = new State();
	System.err.println("[OBProxy: connecting to "+outboard_host+":"+outboard_port+"]");
	ob = connectOutboard(outboard_host, outboard_port);
        new ServiceReplica(id, this, this);
    }
    public OBProxy(int id) {
	st = new State();
	System.err.println("[OBProxy: no outboard]");
	ob = null;
        new ServiceReplica(id, this, this);
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: OBProxy <server id>");
            System.exit(0);
        }

	int index = Integer.parseInt(args[0]);
	if (!(args.length == 2 && args[1].equals("inboard"))) {
	    HostsConfig hosts = new HostsConfig("config", "hosts.config");
	    int port = hosts.getPort(index);
	    new OBProxy(index, "127.0.0.1", port+9);
	}
	else {
	    new OBProxy(index);
	}
    }

    private static byte[] unwrap_bb(ByteBuffer bb) {
	byte[] b = new byte[bb.remaining()];
	bb.get(b);
	return b;
    }
    

    @Override
    public byte[][] appExecuteBatch(byte[][] command, MessageContext[] mcs) {
	try {

	    State st2 = st.copy();
	    byte[][] replies = new byte[command.length][];
	    for (int i = 0; i < command.length; i++) {
		replies[i] = st2.executeSingle(command[i], mcs[i]);
	    }
	    if (null == ob) {
		st = st2;
		return replies;
	    }

	    ByteBuffer ob_stBefore = ByteBuffer.wrap(st.serialize());
	    ByteBuffer ob_stAfter = ByteBuffer.wrap(st2.serialize());
	    assert(command.length == mcs.length);
	    ArrayList<ByteBuffer> ob_command = new ArrayList<ByteBuffer>(command.length);
	    assert (ob_command.size() == command.length);
	    ArrayList<Integer> ob_id = new ArrayList<Integer>(command.length);
	    assert (ob_id.size() == command.length);
	    for(int i = 0 ; i < command.length; i++) {
		ob_command.add(ByteBuffer.wrap(command[i]));
		ob_id.add(new Integer(mcs[i].getConsensusId()));
	    }
	    List<ByteBuffer> ob_replies = ob.appExecuteBatch(ob_stBefore, ob_stAfter, ob_command, ob_id);
	    assert(ob_replies.size() == command.length);
	    replies = new byte[command.length][];
	    for(int i = 0 ; i < command.length; i++) {
		replies[i] = unwrap_bb(ob_replies.get(i));
	    }
	    st = st2;
	    return replies;
	} catch (TException x) {
	    x.printStackTrace();
	    System.exit(-1);
	    return null;
	}
    }

    @Override
    public byte[] appExecuteUnordered(byte[] command, MessageContext msgCtx) {
	System.out.println("state: " + AeSimpleSHA256.convertToHex(st.serialize()));
	if (null == ob) return st.serialize();
	try{
	    ByteBuffer ob_stBefore = ByteBuffer.wrap(st.serialize());
	    ByteBuffer ob_reply = ob.appExecuteUnordered(ob_stBefore, ByteBuffer.wrap(command)) ;

	    return unwrap_bb(ob_reply);
	} catch (TException x) {
	    x.printStackTrace();
	    System.exit(-1);
	    return null;
	}
    }

    @Override
    public void installSnapshot(byte[] state) {
	try{
	    State st2 = new State();
	    st2.deserialize(state);
	    if (null == ob) {
		st = st2 ;
		return ;
	    }
	    ob.installSnapshot(ByteBuffer.wrap(state));
	    st = st2;
	} catch (TException x) {
	    x.printStackTrace();
	    System.exit(-1);
	}
    }

    @Override
    public byte[] getSnapshot() {
	return st.serialize();
    }
}

**Comment by snakejerusalem** [comment]
PS: Another parmeter of interest is "system.totalordermulticast.highMark". If a replica detects that it is still processing consensus X but it is receiving messages of consensus (X + highMark), it will automatically trigger the state transfer.

**Comment by chetmurthy** [comment]
re: highMark, if I understand this riight, once a replica is behind, if I push thru highMark more messages, then the replica will initiate state-transfer, yes?

**Comment by snakejerusalem** [comment]
Correct.

**Comment by snakejerusalem** [comment]
Have you tried my suggestion about decreasing the command rate and increasing the checkpoint period?

**Comment by chetmurthy** [comment]
This doesn't seem to work either:

(1) start 4 replicas
(2) start 1m message writer
(3) stop replica 3
(4) after waiting for at least 10k messages to pass by , resume replica 3

replica 3 never catches up.  I then killed the writer, and found that I could invoke writers with 1k, 10k requests a few times, and then when I did it again, it hung.  Two replicas (3 (resumed) and 1 (continuously running) both had queue-full problems.

Can you point to some description of fault-secenarios that are known to work, that I could try out?

**Comment by snakejerusalem** [comment]
The demos included in the library all seem to work when a replica need to catch-up. Try using the Counter demo. The README file has instructions about how to run it.

**Comment by chetmurthy** [comment]
I copied the Counter example and modified it to add a 1K byte "trailer" to each message from the client.  The server is unmodified.  I get the same sorts of problems.

@@ -33,6 +34,7 @@
  *
  */
 public class CounterClient {
+    public final static int COMMAND_LENGTH = 1024;
 
     public static void main(String[] args) throws IOException {
         if (args.length < 2) {
@@ -43,6 +45,7 @@
         }
 
         ServiceProxy counterProxy = new ServiceProxy(Integer.parseInt(args[0]));
+       Random rand = new Random();
 
         //counterProxy.setInvokeTimeout(1);
 
@@ -80,8 +83,13 @@
                 }
 
                 ByteArrayOutputStream out = new ByteArrayOutputStream(4);
-                new DataOutputStream(out).writeInt(inc);
-
+               DataOutputStream dos = new DataOutputStream(out);
+               dos.writeInt(inc);
+               {
+                   byte[] command = new byte[COMMAND_LENGTH];
+                   rand.nextBytes(command);
+                   dos.write(command);
+               }
                 System.out.println("Counter sending: " + i);
                 byte[] reply;
                 if(inc == 0)

**Comment by chetmurthy** [comment]
BTW, I haven't tried decreasing command-rate, b/c I'm intentionally trying to cause state-transfer.  I can easily believe that a lagging replica will catch up if other replicas have not discarded log-records.  But if state-transfer doesn't work, then ... well, it's not really very useful, is it?

**Comment by snakejerusalem** [comment]
If you can force the Counter demo to reproduce the same problem by adding a trailer to the byte array, that can help us reproducing the bug ourselves. Post here the code of your customized CounterClient (or edit the above post so that the changes you did become more legible), a set of instructions about how to deploy the system and reproduce the behaviour, and the output you get when the problem manifests.

**Comment by chetmurthy** [comment]
Here's the full counter example.  I run it by running

  % ./RUN counter.CounterServer 0 # 1,2,3

to run the client

 %  ./RUN counter.CounterClient 1001 1 1000

etc, as usual.

To elicit the problem, I ran

% ./RUN counter.CounterClient 1001 1 1000000



[counter-example.tar.gz](https://github.com/bft-smart/library/files/273349/counter-example.tar.gz)

**Comment by chetmurthy** [comment]
Oops, I'll post that tarball again, this time with my config files.
[counter-example.tar.gz](https://github.com/bft-smart/library/files/273356/counter-example.tar.gz)

**Comment by snakejerusalem** [comment]
Your RUN script does not execute here. It uses variables of your own, and the paths are not all local to the folder. Please, take the time to prepare a script that is independent from your workstation.

**Comment by snakejerusalem** [comment]
We copy-pasted your code for the Counter demo directly into bft-smart and used your configuration file. We think we successfully reproduced the issue. The problem is the fact that you are reducing the queue size of the communication system in the configuration file; you will need to use the values that come by default. The library is  designed to trigger the state transfer protocol when (1) it detects that is crashed, (2) it detects that it is lagging behind by X number of consensus instances. However, it is not designed to trigger that protocol when the queues get full.

**Comment by chetmurthy** [comment]
Hi,

Thank you for letting me know about the repro.

> The library is is not designed to trigger that [state-change]
> protocol when the queues get full.

The problem with increasing the queue-size, is that my expected
message-sizes vary from a kilobyte, up to several megabytes.  Also, a
system that cannot recover from full queues that cause dropped
messages .... and hangs completely .... isn't really fault-tolerant,
yes?

Cheers,
--chet--

**Comment by snakejerusalem** [comment]
> The problem with increasing the queue-size, is that my expected message-sizes vary from a kilobyte, up to several megabytes.

The queue size is limited by number of messages, not by memory. However, most messages exchanged among replicas will not require a lot of memory. The exception is the PROPOSE message sent by the leader replica (usually replica 0) which includes a batch of client commands.

> Also, a system that cannot recover from full queues that cause dropped messages .... and hangs completely .... isn't really fault-tolerant, yes?

Good observation. BFT-SMaRt implements a modular BFT state machine replication (SMR) protocol which was proven correct under the assumption of reliable channels (respective publications are available [here](http://www.di.fc.ul.pt/~bessani/publications/edcc12-modsmart.pdf) and [here](https://www.zurich.ibm.com/~cca/papers/pax.pdf)), which is a common assumption made by many other BFT-SMR protocols proposed in the literature. We simulate these channels using Java sockets and message queues, but we keep those queues bounded to avoid thrashing. We designed the library to trigger its state transfer if the replica detects that it is delayed by more than X consensus instances, and this requires the assumption that the queues are large enough to store the amount of messages exchanged by each instance. If the queues are not large enough, we run into the issue at hand.

Because in most recent versions of the library we focused much more on making the consensus execution and state transfer mechanism more robust - using a large bound on the queues - we did not explored this issue in much detail. In the future we will stress-test the library with lower bounds and see what we can do to mitigate this problem (probably one of the solutions will be to trigger the state transfer upon this phenomenon, but that won't be enough if more than _f_ replicas exhaust their queues).
